### DISCUSSION TITLE
NDI Support

### DISCUSSION THREAD

**Discussion by esraell** [savonet/liquidsoap#4165] [Ideas]
Hi,

Would it be possible to add NDI support to Liquidsoap for sending and receiving NDI streams? This feature would be incredibly useful in a studio or any local environment. It would greatly enhance flexibility in handling video and audio streams within such setups. I’d love to see if this could be considered or explored for future releases.

Thank you for your time and consideration!

**Comment by vitoyucepi** [comment]
Hi @esraell,
Could you please explain what NDI is and how to get started with it?
Are there any reference implementations or specifications? Do you know anything about open source implementations?

**Reply by esraell** [reply]
Thank you for your response! NDI is a protocol developed by NewTek that enables high-quality video and audio transmission over IP networks, making it ideal for studio setups and local environments where low latency is important.

For more information on implementing NDI, here's a useful link: [NDI for Developers](https://ndi.video/for-developers/).

**Reply by toots** [reply]
Thanks for these links! What kind of application would you have in mind using it @esraell ? Do you know any tool to test our implementation?

**Reply by esraell** [reply]
Hi @toots,

Thanks for following up! The primary application I have in mind would be for live studio environments where video and audio are streamed across multiple devices in real-time, such as broadcasting setups or live events. NDI would allow Liquidsoap to become a central hub for both audio and video processing over IP networks, significantly enhancing its flexibility in such scenarios.

For testing the implementation, tools like NDI Studio Monitor and NDI Test Patterns from NewTek’s NDI Tools package are commonly used. They allow you to send and receive NDI streams and monitor performance. Additionally, there are some open-source libraries like [OBS-NDI](https://github.com/DistroAV/DistroAV) that might provide more context on how others have implemented NDI in open-source projects.

Let me know your thoughts!

**Reply by toots** [reply]
Thanks! I have worked on this. The API is really straight forward and the protocol seems widely used, in particular for hardware stuff.

I've got an output working already. I'm hoping to get a basic experimental input and output added for to the upcoming `2.3.0` release and then work on a more extensive support after it.

**Comment by toots** [comment]
First experimental output is here: https://github.com/savonet/liquidsoap/pull/4181 will have to do the rest after the release.

**Comment by esraell** [comment]
@toots Thank you for implementing NDI and considering this idea. I’d like to inform you that they’ve added the [NDI Free Audio](https://docs.ndi.video/guides/guides/utilities/ndi-free-audio) tool, which enables the transmission and reception of audio over NDI via the Command Line. Thank you for your work.
